src/core/qgsauxiliarystorage.cpp

```diff
@@ -0,0 +1,659 @@
+}
+
+//
+// QgsAuxiliaryLayer
+//
+
+QgsAuxiliaryLayer::QgsAuxiliaryLayer( const QString &pkField, const QString &filename, const QString &table, const QgsVectorLayer *vlayer )
+  : QgsVectorLayer( QString( "%1|layername=%2" ).arg( filename, table ), QString( "%1_auxiliarystorage" ).arg( table ), "ogr" )
+  , mLayer( vlayer )
+{
+  // init join info
+  mJoinInfo.setPrefix( AS_JOINPREFIX );
+  mJoinInfo.setJoinLayer( this );
+  mJoinInfo.setJoinFieldName( AS_JOINFIELD );
+  mJoinInfo.setTargetFieldName( pkField );
+  mJoinInfo.setEditable( true );
+  mJoinInfo.setUpsertOnEdit( true );
+  mJoinInfo.setCascadedDelete( true );
+  mJoinInfo.setJoinFieldNamesBlackList( QStringList() << QStringLiteral( "rowid" ) ); // introduced by ogr provider
+}
+
+QgsVectorLayer *QgsAuxiliaryLayer::toSpatialLayer() const
+{
+  QgsVectorLayer *layer = QgsMemoryProviderUtils::createMemoryLayer( QStringLiteral( "auxiliary_layer" ), fields(), mLayer->wkbType(), mLayer->crs() );
+
+  QString pkField = mJoinInfo.targetFieldName();
+  QgsFeature joinFeature;
+  QgsFeature targetFeature;
+  QgsFeatureIterator it = getFeatures();
+
+  layer->startEditing();
+  while ( it.nextFeature( joinFeature ) )
+  {
+    QString filter = QgsExpression::createFieldEqualityExpression( pkField, joinFeature.attribute( AS_JOINFIELD ) );
+
+    QgsFeatureRequest request;
+    request.setFilterExpression( filter );
+
+    mLayer->getFeatures( request ).nextFeature( targetFeature );
+
+    if ( targetFeature.isValid() )
+    {
+      QgsFeature newFeature( joinFeature );
+      newFeature.setGeometry( targetFeature.geometry() );
+      layer->addFeature( newFeature );
+    }
+  }
+  layer->commitChanges();
+
+  return layer;
+}
+
+QgsVectorLayerJoinInfo QgsAuxiliaryLayer::joinInfo() const
+{
+  return mJoinInfo;
+}
+
+bool QgsAuxiliaryLayer::exists( const QgsPropertyDefinition &definition ) const
+{
+  return ( fields().indexOf( QgsAuxiliaryField::nameFromProperty( definition ) ) >= 0 );
+}
+
+bool QgsAuxiliaryLayer::addAuxiliaryField( const QgsPropertyDefinition &definition )
+{
+  if ( definition.name().isEmpty() || exists( definition ) )
+    return false;
+
+  const QgsAuxiliaryField af( definition );
+  const bool rc = addAttribute( af );
+  updateFields();
+
+  return rc;
+}
+
+QgsAuxiliaryFields QgsAuxiliaryLayer::auxiliaryFields() const
+{
+  QgsAuxiliaryFields afields;
+
+  for ( int i = 2; i < fields().count(); i++ ) // ignore rowid and PK field
+    afields.append( QgsAuxiliaryField( fields().field( i ) ) );
+
+  return afields;
+}
+
+bool QgsAuxiliaryLayer::deleteAttribute( int attr )
+{
+  QgsVectorLayer::deleteAttribute( attr );
+  bool rc = commitChanges();
+  startEditing();
+  return rc;
+}
+
+bool QgsAuxiliaryLayer::save()
+{
+  bool rc = false;
+
+  if ( isEditable() )
+  {
+    rc = commitChanges();
+  }
+
+  startEditing();
+
+  return rc;
+}
+
+int QgsAuxiliaryLayer::createProperty( QgsPalLayerSettings::Property property, const QString &providerId, QgsVectorLayer *layer )
```

Comment:
I think this specific use method belongs in QgsPalLayerSettings, and the diagram version in QgsDiagramLayerSettings. It seems messy to me to have usage specific functions in QgsAuxiliaryLayer.

Reply:
Are you sure? Because it's pretty specific to the auxiliary storage stuff considering that the aim is to create a new field in the auxiliary layer. However it's true that the new property is then activated in layer settings... Maybe the method is badly named and leads to confusion?